Add support for enabling extended result codes as default on a connection

intolerable opened the pull request:
- Added support for the `SQLITE_OPEN_EXRESCODE` `sqlite3_open_v2` flag, which enables by-default extended result codes on connections it opens (and also causes `sqlite3_open_v2` itself to return extended result codes)
- Added the `sqlite3_extended_result_codes` function to switch between primary and extended result codes on an existing connection

There's one thing that may need to be changed before this can be merged: with extended result codes enabled for a given connection, there are now two possible "success" results (not counting `SQLITE_ROW` and `SQLITE_DONE`) -- `SQLITE_OK` and `SQLITE_OK_LOAD_PERMANENTLY` -- while `Database.SQLite3.Direct.toResult`(`M`) considers anything not exactly `SQLITE_OK` to be an error. This shouldn't be a problem unless this library adds support for `sqlite3_load_extension`, though (https://www.sqlite.org/rescode.html#ok_load_permanently), and should be a simple fix (`toResult` can mask off the 8 least significant bits before comparing to 0).

intolerable commented on Database/SQLite3/Direct.hs:
```diff
@@ -314,6 +315,11 @@ extendedErrcode :: Database -> IO Error
+-- | <https://www.sqlite.org/c3ref/extended_result_codes.html>
+setExtendedResultCodes :: Database -> Bool -> IO Error
+setExtendedResultCodes (Database db) enabled =
+    decodeError <$> c_sqlite3_extended_result_codes db enabled
```
good point, fixed 👍

jchia reviewed (changes requested):
There is only remaining code issue. After this, I'd like to combine all the commits into one, especially those responding to my review. If you can't do it, I'll do the git history manipulations myself.

intolerable commented:
I've added an explicit check for the error, and squashed the branch. do you think anything should be done at this point to handle the `OK_LOAD_PERMANENTLY` case in `toResult`? if not, this branch should be mergeable -- thank you for your feedback!

jchia commented:
The commit message is way too long. I'm going to change the commit message, commit and push.
